src/CaptureFile/CaptureFile.cpp

```diff
@@ -0,0 +1,189 @@
+    return ErrorMessage{
+        absl::StrFormat("Incompatible version %d, expected %d", header_.version, kFileVersion)};
+  }
+
+  if (header_.section_list_offset == 0) {
+    return outcome::success();
+  }
+
+  OUTCOME_TRY(number_of_sections,
+              orbit_base::ReadFullyAtOffset<uint16_t>(fd_, header_.section_list_offset));
+  std::vector<CaptureFileSection> section_list{number_of_sections};
+  OUTCOME_TRY(bytes_read,
+              orbit_base::ReadFullyAtOffset(fd_, section_list.data(),
+                                            number_of_sections * sizeof(CaptureFileSection),
+                                            header_.section_list_offset + sizeof(uint16_t)));
+
+  if (bytes_read < number_of_sections * sizeof(CaptureFileSection)) {
+    return ErrorMessage{absl::StrFormat(
+        "Unexpected EOF while reading section list: section list size=%d, bytes available=%d",
+        number_of_sections * sizeof(CaptureFileSection), bytes_read)};
+  }
+
+  section_list_ = std::move(section_list);
+
+  return outcome::success();
+}
+
+const std::vector<CaptureFileSection>& CaptureFileImpl::GetSectionList() const {
+  return section_list_;
+}
+
+ErrorMessageOr<void> orbit_capture_file::CaptureFileImpl::WriteToSection(uint64_t section_number,
+                                                                         uint64_t offset_in_section,
+                                                                         const void* data,
+                                                                         size_t size) {
+  CHECK(section_number < section_list_.size());
+
+  const CaptureFileSection& section = section_list_[section_number];
+  CHECK(offset_in_section + size <= section.size);
+
+  OUTCOME_TRY(orbit_base::WriteFullyAtOffset(fd_, data, size, section.offset + offset_in_section));
+
+  return outcome::success();
+}
+
+template <uint64_t alignment>
+constexpr uint64_t AlignUp(uint64_t value) {
+  // alignment must be a power of 2
+  static_assert((alignment & (alignment - 1)) == 0);
+  return (value + (alignment - 1)) & ~(alignment - 1);
+}
+
+ErrorMessageOr<uint16_t> orbit_capture_file::CaptureFileImpl::AddSection(uint64_t section_type,
+                                                                         uint64_t section_size) {
+  if (section_list_.size() == std::numeric_limits<uint16_t>::max()) {
+    return ErrorMessage{
+        absl::StrFormat("Section list has reached its maximum size: %d", section_list_.size())};
+  }
+
+  // Take a copy of section list
+  auto section_list = section_list_;
+
+  // The current file format always places the section list is at the end of the file
+  // therefore the current section_offset is going to be the offset for the new section.
+  section_list.push_back(CaptureFileSection{/*.type = */ section_type,
+                                            /*.offset = */ header_.section_list_offset,
+                                            /*.size = */ section_size});
```

Comment:
well, this is quite a bug... not sure why tests did not catch it - converted to draft. will need another review probably tomorrow.

Reply:
fixed(implementation and the test)